storage: ensure leaseholder gossips system config

spencerkimball opened the pull request:
We've experienced a recurrence of the system config not being gossiped
at cluster startup non-deterministically. This change more directly
addresses the issue of whether or not the leaseholder should gossip by
checking whether the gossiped version of the system config (or node
liveness) differs from the canonical data, and gossiping if so.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13011)
<!-- Reviewable:end -->

spencerkimball committed:
storage: ensure leaseholder gossips system config

We've experienced a recurrence of the system config not being gossiped
at cluster startup non-deterministically. This change more directly
addresses the issue of whether or not the leaseholder should gossip by
checking whether the gossiped version of the system config (or node
liveness) differs from the canonical data, and gossiping if so.

petermattis commented on pkg/storage/replica.go:
```diff
@@ -252,11 +252,15 @@ type Replica struct {
-	abortCache   *AbortCache // Avoids anomalous reads after abort
+
+	// sha1 hash of the system config @ last gossip. Must be accessed only
+	// from the Raft-processing goroutine.
```
There is no `Raft-processing goroutine` anymore. What's the real locking story here?

petermattis commented on pkg/storage/replica.go:
```diff
@@ -252,11 +252,15 @@ type Replica struct {
+	systemConfigHash []byte
+	// sha1 hash of node livenesses @ last gossip. Must be accessed only
+	// from the Raft-processing goroutine.
+	nodeLivenessHash []byte
```
Given that both the system config span and node liveness span should exist in only a single range, I think we could move these fields to `Store`. Not terribly important, but it is somewhat weird that they exist here as only a single range in the system will need them.

Btw, do we ensure the node liveness span is not split as we do with the system config span?

petermattis commented on pkg/storage/node_liveness.go:
```diff
@@ -67,6 +69,32 @@ var (
+
+// Hash returns a SHA1 hash of the contents.
+func (ls livenessSlice) Hash() []byte {
+	sort.Sort(ls)
```
Feels sort of strange that `Hash` has the side-effect of `Sorting`. Perhaps this should be inlined in `NodeLiveness.Hash`.

petermattis reviewed (approved):
LGTM

tamird commented:
This is largely untested; can you add a test that exercises the new gossip condition?

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/node_liveness.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kasj2HZScQ6qKld55_V:-Kasj2H_N4YWbTH7lCTI:bh9eme9) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L85)):*
> ```Go
> 
> // Hash returns a SHA1 hash of the contents.
> func (ls livenessSlice) Hash() []byte {
> ```

can you plumb a slice through this? in the ultimate caller, you can say `hash = foo.Hash(hash[:0])` to save some allocs

---

*[pkg/storage/node_liveness.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasiRm5nhJ75SmXQuU9:-KasiRm5nhJ75SmXQuUA:bbckrk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L93)):*
> ```Go
> 			panic(err.Error())
> 		}
> 		_, _ = sha.Write(data)
> ```

already panicking above, do the same here?

---

*[pkg/storage/replica.go, line 4386 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjBjVHdlgBnAu_70y:-KasjBjVHdlgBnAu_70z:bk9grsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4386)):*
> ```Go
> 		ls[i] = liveness
> 
> 		if bytes.Compare(kv.Key, span.Key) >= 0 && bytes.Compare(kv.Key, span.EndKey) < 0 {
> ```

this is unrelated to the rest of this change, correct? why is it necessary?

---

*[pkg/storage/replica_proposal.go, line 734 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjjmDqVjR75TqFGSE:-KasjjmEJp8B1CntYv9j:bdi2n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e76916624868a64fb779ac6b9d19fefe462bf6/pkg/storage/replica_proposal.go#L734)):*
> ```Go
> 			}
> 		}
> 		if lResult.maybeGossipNodeLiveness != nil {
> ```

why the move?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-:-KasjmZMpOM6oqmklSsg:b-687dl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'm not sure this will solve all our problems, but it should solve at least some of them.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/storage/replica.go, line 4317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KaslwO8mRh_iUtTEQVv:-KaslwO9nUvy_mfBdyie:b-az9epq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4317)):*
> ```Go
> 	}
> 
> 	// If the contents in gossip match last read system config, skip gossip.
> ```

`r.systemConfigHash` isn't really the "last read system gossip" though, is it? We only update it to a non-nil value after successfully gossiping it. Not that the code doesn't look like it'll work, but the comment is a bit misleading.

---

*[pkg/storage/replica.go, line 4357 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasmWNyizzzPXBxI3y0:-KasmWNyizzzPXBxI3y1:b-ijvgr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4357)):*
> ```Go
> 	}
> 
> 	// If gossiped node liveness matches last read node liveness records, skip gossip.
> ```

Same comment as about `r.systemConfigHash`

---

*[pkg/storage/replica.go, line 4386 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjBjVHdlgBnAu_70y:-KasnE-BWj_r_nmaa3r2:b-civfxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4386)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is unrelated to the rest of this change, correct? why is it necessary?
</blockquote></details>

It looks necessary due to the above change from scanning just the provided span to scanning the entire `keys.NodeLivenessSpan`

---

*[pkg/storage/replica_proposal.go, line 439 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasnPBoDsrkskJfx7lS:-KasnPBoDsrkskJfx7lT:b-wfjfkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e76916624868a64fb779ac6b9d19fefe462bf6/pkg/storage/replica_proposal.go#L439)):*
> ```Go
> 	// gossiped.
> 	if iAmTheLeaseHolder {
> 		if leaseChangingHands || atomic.CompareAndSwapInt32(&r.store.haveGossipedSystemConfig, 0, 1) {
> ```

Thanks for getting rid of this atomic logic 👍 -- it didn't account for any sort of errors in attempting to gossip the system config

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-:-Kaso7Ji1iYzWy6NMTWm:b-6urar)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/storage/node_liveness.go, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatwVhSn7s4Nfa5_cSC:-KatwVhSn7s4Nfa5_cSD:b-xwbons) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L20)):*
> ```Go
> 
> import (
> 	"crypto/sha1"
> ```

Why sha1? There's no good reason to use it instead of sha256. 

---

*[pkg/storage/node_liveness.go, line 344 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatwycycLNxe5pMyW2j:-KatwycycLNxe5pMyW2k:b1fgclv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L344)):*
> ```Go
> }
> 
> // Hash returns a SHA1 hash of the NodeLiveness contents.
> ```

Don't mention the hash algorithm in the comments; it's an implementation detail.

---

*[pkg/storage/replica.go, line 4318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatyVsmWYPXJkJ-3w1f:-KatyVsmWYPXJkJ-3w1g:b-htkj43) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4318)):*
> ```Go
> 
> 	// If the contents in gossip match last read system config, skip gossip.
> 	if cfg, ok := r.store.Gossip().GetSystemConfig(); ok && bytes.Equal(cfg.Hash(), r.systemConfigHash) {
> ```

Doesn't reading this from gossip defeat the purpose of this code? The config in gossip should always match the last one we gossiped, even if the one we would load on the next line is different. 

---

*[pkg/storage/replica.go, line 4386 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjBjVHdlgBnAu_70y:-KatzCTgqs7qRfoy4U1c:b7d3qbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4386)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It looks necessary due to the above change from scanning just the provided span to scanning the entire `keys.NodeLivenessSpan`
</blockquote></details>

This needs some comments: We read the entire liveness span so we can compute the correct hash, even though we only gossip entries that fall within the requested span. 

---

*[pkg/storage/replica_proposal.go, line 440 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kau-38VEHvS4hz2wcyE:-Kau-38VEHvS4hz2wcyF:b-inmvuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica_proposal.go#L440)):*
> ```Go
> 		if leaseChangingHands {
> 			r.systemConfigHash = nil
> 			r.nodeLivenessHash = nil
> ```

This seems insufficient. The next call to `maybeGossipNodeLiveness` will see the nil hash and gossip only the requested `span`, then set the hash suggesting that the entire span has been gossiped. Instead, it needs to gossip the entire liveness span if `r.nodeLivenessHash` is nil. 

This whole change feels like a much *less* direct approach than the old atomic booleans to me. I don't see exactly what was going wrong with the booleans so it's difficult for me to feel confident that this will fix any problems (It seems to me like one of the most likely problems with the booleans is that they treat the entire liveness span as a single unit when we often operate on a subset of it. Hashing it would still have the same problem). This logic really needs some tests. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Alex, what other problems did you have in mind?

---

Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/storage/node_liveness.go, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatwVhSn7s4Nfa5_cSC:-Kawm4_wDZGJgksU4T98:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L20)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why sha1? There's no good reason to use it instead of sha256. 
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kasj2HZScQ6qKld55_V:-KawdoLwM_wkXC8kg_OK:b-tu9oyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you plumb a slice through this? in the ultimate caller, you can say `hash = foo.Hash(hash[:0])` to save some allocs
</blockquote></details>

I'm going to leave as is because this is how the system config API works. There's no real need to save the `[]byte` allocation here.

---

*[pkg/storage/node_liveness.go, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasiFiedFHzLuUwXFsu-r1-86:-KawepYL6qhuF8mJ-qHt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L86)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Feels sort of strange that `Hash` has the side-effect of `Sorting`. Perhaps this should be inlined in `NodeLiveness.Hash`.
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasiRm5nhJ75SmXQuU9:-Kaweq4Bw2i7oluB0oyv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L93)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

already panicking above, do the same here?
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 344 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatwycycLNxe5pMyW2j:-KawerDy-U2WcIokdMGL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/node_liveness.go#L344)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Don't mention the hash algorithm in the comments; it's an implementation detail.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasiFiedFHzLuUwXFsw-r1-257:-Kawfhv9UIdmUWKiQ2qq:bk637v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L257)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is no `Raft-processing goroutine` anymore. What's the real locking story here?
</blockquote></details>

Moved to store and am using `atomic.Value` to give it a real locking story.

---

*[pkg/storage/replica.go, line 261 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasiFiedFHzLuUwXFsw-r1-261:-KawgRIJ4zKSZBGOEAOz:b45i3cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L261)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Given that both the system config span and node liveness span should exist in only a single range, I think we could move these fields to `Store`. Not terribly important, but it is somewhat weird that they exist here as only a single range in the system will need them.

Btw, do we ensure the node liveness span is not split as we do with the system config span?
</blockquote></details>

Moved to store. Yes, it's prevented from being split. See `pkg/keys/spans.go` in `NoSplitSpans`.

---

*[pkg/storage/replica.go, line 4317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KaslwO8mRh_iUtTEQVv:-KawjUuLIXa3z_duQv9s:b-i5phlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4317)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

`r.systemConfigHash` isn't really the "last read system gossip" though, is it? We only update it to a non-nil value after successfully gossiping it. Not that the code doesn't look like it'll work, but the comment is a bit misleading.
</blockquote></details>

I'm now updating the hash before we attempt to gossip. And the comment is "last read system config", not "last read system gossip", so I think it makes sense now.

---

*[pkg/storage/replica.go, line 4318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatyVsmWYPXJkJ-3w1f:-KawjfW-HgcFfL_rG5I6:bnknqnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4318)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't reading this from gossip defeat the purpose of this code? The config in gossip should always match the last one we gossiped, even if the one we would load on the next line is different. 
</blockquote></details>

The problem is if another replica gossiped or the gossip is out of date. Gossip also relies on wall clock times and so can have strange things happen when multiple nodes gossip the same key. This code ensures that we do gossip the latest value.

---

*[pkg/storage/replica.go, line 4357 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasmWNyizzzPXBxI3y0:-KawlYDo7TAmVjl4adjU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4357)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Same comment as about `r.systemConfigHash`
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4386 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjBjVHdlgBnAu_70y:-KawlkBZPbz29M1xZd9d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4386)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This needs some comments: We read the entire liveness span so we can compute the correct hash, even though we only gossip entries that fall within the requested span. 
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 439 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasnPBoDsrkskJfx7lS:-KayJfPxV6zIy6Kn5waK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e76916624868a64fb779ac6b9d19fefe462bf6/pkg/storage/replica_proposal.go#L439)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Thanks for getting rid of this atomic logic 👍 -- it didn't account for any sort of errors in attempting to gossip the system config
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 440 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kau-38VEHvS4hz2wcyE:-KayIKFZ-i0DmJaqu-ON:bgmr8sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica_proposal.go#L440)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems insufficient. The next call to `maybeGossipNodeLiveness` will see the nil hash and gossip only the requested `span`, then set the hash suggesting that the entire span has been gossiped. Instead, it needs to gossip the entire liveness span if `r.nodeLivenessHash` is nil. 

This whole change feels like a much *less* direct approach than the old atomic booleans to me. I don't see exactly what was going wrong with the booleans so it's difficult for me to feel confident that this will fix any problems (It seems to me like one of the most likely problems with the booleans is that they treat the entire liveness span as a single unit when we often operate on a subset of it. Hashing it would still have the same problem). This logic really needs some tests. 
</blockquote></details>

I think you are right; this original change was insufficient for node liveness. I've changed things so we don't gossip on lease acquisition, but instead more directly from the gossip loop in `Store.startGossip`.

There was logic in that method which was incorrect around the wait group for store init completion. We also now have a retry loop to ensure that system data is gossiped quickly (and doesn't wait for the 60s interval before trying again) in the event that any attempt (especially the initial attempt) fails.

---

*[pkg/storage/replica_proposal.go, line 734 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjjmDqVjR75TqFGSE:-KawlxEdhhOr8twtLNNA:b180jmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e76916624868a64fb779ac6b9d19fefe462bf6/pkg/storage/replica_proposal.go#L734)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the move?
</blockquote></details>

More consistent with gossiping the system config.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-:-KayJaZErhZ4gqPHVH-R:bce4tpt)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 8 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/storage/node_liveness.go, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kb6WqkLN_ym5hm3SE1v:-Kb6WqkM82BJREmADI6M:bbt0s2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/node_liveness.go#L91)):*
> ```Go
> 		data, err := l.Marshal()
> 		if err != nil {
> 			panic(err.Error())
> ```

nit here and below: `.Error()` is not needed

---

*[pkg/storage/store.go, line 1287 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kb6XEEpVgMS5mxOMefp:-Kb6XEEpVgMS5mxOMefq:b-om42vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/store.go#L1287)):*
> ```Go
> 		{
> 			fn: func(ctx context.Context) error {
> 				return s.maybeGossipSystemData(ctx, keys.SystemConfigSpan, func(r *Replica) error {
> ```

why is the signature `(*Replica) error`? if you made it `(context.Context) error` you could just pass `r.maybeGossipSystemConfig` here and `r.maybeGossipNodeLiveness` below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 4317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KaslwO8mRh_iUtTEQVv:-KbB5dVq-mjjUL-gjlGR:b-8ylp80) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4317)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I'm now updating the hash before we attempt to gossip. And the comment is "last read system config", not "last read system gossip", so I think it makes sense now.
</blockquote></details>

s/last read/last gossiped/? I think a comment that `systemConfigHash` is updated after successful gossip and cleared when there is any change would be helpful.

---

*[pkg/storage/replica.go, line 4318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KatyVsmWYPXJkJ-3w1f:-KbB5nif36Y8UWsVgQnq:b-t7cxgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4318)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

The problem is if another replica gossiped or the gossip is out of date. Gossip also relies on wall clock times and so can have strange things happen when multiple nodes gossip the same key. This code ensures that we do gossip the latest value.
</blockquote></details>

OK, please add a comment that this is to detect gossip from other nodes and other weirdness. 

---

*[pkg/storage/replica.go, line 4371 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbB8D-NHfzh_5X5lykp:-KbB8D-NHfzh_5X5lykq:b-vtqhdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/replica.go#L4371)):*
> ```Go
> 	kvs := br.Responses[0].GetInner().(*roachpb.ScanResponse).Rows
> 
> 	// Gossip only node liveness records within the requested span.
> ```

There is no longer a requested span. 

---

*[pkg/storage/replica_proposal.go, line 708 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbB8L7xnCv8wNBquQtL:-KbB8L7xnCv8wNBquQtM:b-l9fcow) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/replica_proposal.go#L708)):*
> ```Go
> 	if lResult.maybeGossipNodeLiveness != nil {
> 		r.store.nodeLivenessHash.Store([]byte(nil))
> 		if err := r.maybeGossipNodeLiveness(ctx); err != nil {
> ```

The `Span` value of `maybeGossipNodeLiveness` is no longer used; we re-gossip everything when any value changes. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 1328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbBGON9TWSZWmbPflWw:-KbBGON9TWSZWmbPflWx:b-jpj2bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/store.go#L1328)):*
> ```Go
> 					s.initComplete.Done()
> 					first = false
> 					if s.cfg.TestingKnobs.DisablePeriodicGossips {
> ```

Moving this knob here will likely cause a bunch of test flakiness. The intention when I added it was to never perform the periodic gossips. What caused you to need to move it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2593 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbBj2hWELWNqLppbNc8:-KbBj2hWELWNqLppbNc9:be6ah1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/client_raft_test.go#L2593)):*
> ```Go
> 	replica := mtc.stores[0].LookupReplica(splitKey, nil)
> 	if replica == nil {
> 		t.Fatalf("lookup replica at key %s returned nil", splitKey)
> ```

nit: %q would be nicer here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
I don't have any other concrete problems to address for this issue beyond what you've done. The biggest issue I had was with the use of the atomic bools that disregarded whether or not the gossiping succeeded, along with a concern about the general edge-triggered (rather than level-triggered) approach to the whole thing. The change to more reliably call the gossip functions satisfies that concern.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-:-KbBnfh4WI9wCz2oAjBc:b-jwkt41)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2593 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbBj2hWELWNqLppbNc8:-KbBopeABrvsq48aR04R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/client_raft_test.go#L2593)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

nit: %q would be nicer here
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KaslwO8mRh_iUtTEQVv:-KbHOJl24bJf2ysHaiwz:b-q54ww9) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4317)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/last read/last gossiped/? I think a comment that `systemConfigHash` is updated after successful gossip and cleared when there is any change would be helpful.
</blockquote></details>

Removed this stuff.

---

*[pkg/storage/replica.go, line 4357 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasmWNyizzzPXBxI3y0:-KbHOLQLWykwCnp1gYOF:b-q1bu4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4357)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This code is gone.

---

*[pkg/storage/replica.go, line 4386 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KasjBjVHdlgBnAu_70y:-KbHOPfB7rIrLPLoMptA:b-q1bu4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica.go#L4386)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This code is gone.

---

*[pkg/storage/replica.go, line 4371 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbB8D-NHfzh_5X5lykp:-KbHONBToG0C7sg2YnN5:bw404uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/replica.go#L4371)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There is no longer a requested span. 
</blockquote></details>

Reverted this bit.

---

*[pkg/storage/replica_proposal.go, line 440 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-Kau-38VEHvS4hz2wcyE:-KbHOVSu4Z2uPEZCtDWN:b-iz48l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae4e693faaa05fb4b1a20d5d59a18b091a65c7e/pkg/storage/replica_proposal.go#L440)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I think you are right; this original change was insufficient for node liveness. I've changed things so we don't gossip on lease acquisition, but instead more directly from the gossip loop in `Store.startGossip`.

There was logic in that method which was incorrect around the wait group for store init completion. We also now have a retry loop to ensure that system data is gossiped quickly (and doesn't wait for the 60s interval before trying again) in the event that any attempt (especially the initial attempt) fails.
</blockquote></details>

I've reverted a bunch of this code; @bdarnell can you take another look?

---

*[pkg/storage/replica_proposal.go, line 708 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbB8L7xnCv8wNBquQtL:-KbHO_XPj7hLohnIuu17:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/replica_proposal.go#L708)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The `Span` value of `maybeGossipNodeLiveness` is no longer used; we re-gossip everything when any value changes. 
</blockquote></details>

Reverted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 5 of 8 files at r4.
Review status: 4 of 6 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 6 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 1328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbBGON9TWSZWmbPflWw:-KbKtJWbchE4ArClkJvg:bep9mzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd6797314d905eadb7f643c37eb8ec2c771291b3/pkg/storage/store.go#L1328)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Moving this knob here will likely cause a bunch of test flakiness. The intention when I added it was to never perform the periodic gossips. What caused you to need to move it?
</blockquote></details>

Ping. Any clue on why this moved? Doing so doesn't seem right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 440 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbLygyazg0vRGu4N9fk:-KbLygyazg0vRGu4N9fl:b-m05pdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ada27e882b4d8a4708e95e6b3ddf45abd7786be/pkg/storage/replica_proposal.go#L440)):*
> ```Go
> 	// or this is the first time that either system data has been
> 	// gossiped.
> 	if iAmTheLeaseHolder {
> ```

I think we still need this to regossip both items if `iAmTheLeaseHolder && leaseChangingHands`. The atomic stuff is no longer needed 

---

*[pkg/storage/replica_test.go, line 986 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011#-KbLzU8R6Ffqz8Biz_0q:-KbLzU8R6Ffqz8Biz_0r:b-8yg54z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ada27e882b4d8a4708e95e6b3ddf45abd7786be/pkg/storage/replica_test.go#L986)):*
> ```Go
> // TestReplicaGossipConfigsOnLease verifies that config info is gossiped
> // upon acquisition of the range lease.
> func TestReplicaGossipConfigsOnLease(t *testing.T) {
> ```

I think this test should be reinstated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13011)*
<!-- Sent from Reviewable.io -->

tamird commented:
This PR contains a bunch of random changes that don't really have anything to do with the problems ostensibly being fixed here.

I've sent (and merged) a series of PRs:
- https://github.com/cockroachdb/cockroach/pull/13125
- https://github.com/cockroachdb/cockroach/pull/13142
- https://github.com/cockroachdb/cockroach/pull/13147
where I've "backported" the unrelated changes into master.

With those changes applied, there following changes remain here:
- `store.haveGossiped{SystemConfig,NodeLiveness}` are replaced with hashes, but that doesn't make any sense, since the full structures are available for comparison and, before this change, already were being compared in the case of node liveness keys, and in the case of the system config are being fully compared as of #13125
- `storage.systemDataGossipInterval` is reduced from 1 minute to 15 seconds
- `(*store).maybeGossipSystemData` actually calls `maybeGossip{SystemConfig,NodeLiveness}`, rather than relying on those functions being called on acquisition of the range lease

I don't see how these changes fix anything. This really needs a failing test case before we can proceed any further.

tamird commented:
For reference, here's my latest version of this PR: https://github.com/tamird/cockroach/commit/system-config-gossip-new
